### ISSUE TITLE
Error in generate statistics due to Python Snappy

### ISSUE THREAD

**Issue by vincentteyssier** [tensorflow/data-validation#21]
Hi, 

I got a fresh install of Ubuntu 16.04 on AWS, installed the following in this order:
- python 2.7
- python-dev
- python-snappy
- tensorflow 1.11
- apache-beam
- jupyter
- tensorflow-data-validation

I manage to import tfdv without issue however the generate statistics fails with the following trace:

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-3-2fbc1f02b217> in <module>()
----> 1 stats = tfdv.generate_statistics_from_csv(DATA, delimiter=';')

/usr/local/lib/python2.7/dist-packages/tensorflow_data_validation/utils/stats_gen_lib.pyc in generate_statistics_from_csv(data_location, column_names, delimiter, output_path, stats_options, pipeline_options)
    153             shard_name_template='',
    154             coder=beam.coders.ProtoCoder(
--> 155                 statistics_pb2.DatasetFeatureStatisticsList)))
    156   return load_statistics(output_path)
    157 

/usr/local/lib/python2.7/dist-packages/apache_beam/pipeline.pyc in __exit__(self, exc_type, exc_val, exc_tb)
    412   def __exit__(self, exc_type, exc_val, exc_tb):
    413     if not exc_type:
--> 414       self.run().wait_until_finish()
    415 
    416   def visit(self, visitor):

/usr/local/lib/python2.7/dist-packages/apache_beam/pipeline.pyc in run(self, test_runner_api)
    392     if test_runner_api and self._verify_runner_api_compatible():
    393       return Pipeline.from_runner_api(
--> 394           self.to_runner_api(), self.runner, self._options).run(False)
    395 
    396     if self._options.view_as(TypeOptions).runtime_type_check:

/usr/local/lib/python2.7/dist-packages/apache_beam/pipeline.pyc in run(self, test_runner_api)
    405       finally:
    406         shutil.rmtree(tmpdir)
--> 407     return self.runner.run_pipeline(self)
    408 
    409   def __enter__(self):

/usr/local/lib/python2.7/dist-packages/apache_beam/runners/direct/direct_runner.pyc in run_pipeline(self, pipeline)
    133       runner = BundleBasedDirectRunner()
    134 
--> 135     return runner.run_pipeline(pipeline)
    136 
    137 

/usr/local/lib/python2.7/dist-packages/apache_beam/runners/portability/fn_api_runner.pyc in run_pipeline(self, pipeline)
    221     from apache_beam.runners.dataflow.dataflow_runner import DataflowRunner
    222     pipeline.visit(DataflowRunner.group_by_key_input_visitor())
--> 223     return self.run_via_runner_api(pipeline.to_runner_api())
    224 
    225   def run_via_runner_api(self, pipeline_proto):

/usr/local/lib/python2.7/dist-packages/apache_beam/runners/portability/fn_api_runner.pyc in run_via_runner_api(self, pipeline_proto)
    224 
    225   def run_via_runner_api(self, pipeline_proto):
--> 226     return self.run_stages(*self.create_stages(pipeline_proto))
    227 
    228   def create_stages(self, pipeline_proto):

/usr/local/lib/python2.7/dist-packages/apache_beam/runners/portability/fn_api_runner.pyc in run_stages(self, pipeline_components, stages, safe_coders)
    862         metrics_by_stage[stage.name] = self.run_stage(
    863             controller, pipeline_components, stage,
--> 864             pcoll_buffers, safe_coders).process_bundle.metrics
    865     finally:
    866       controller.close()

/usr/local/lib/python2.7/dist-packages/apache_beam/runners/portability/fn_api_runner.pyc in run_stage(self, controller, pipeline_components, stage, pcoll_buffers, safe_coders)
    975     return BundleManager(
    976         controller, get_buffer, process_bundle_descriptor,
--> 977         self._progress_frequency).process_bundle(data_input, data_output)
    978 
    979   # These classes are used to interact with the worker.

/usr/local/lib/python2.7/dist-packages/apache_beam/runners/portability/fn_api_runner.pyc in process_bundle(self, inputs, expected_outputs)
   1179         process_bundle=beam_fn_api_pb2.ProcessBundleRequest(
   1180             process_bundle_descriptor_reference=self._bundle_descriptor.id))
-> 1181     result_future = self._controller.control_handler.push(process_bundle)
   1182 
   1183     with ProgressRequester(

/usr/local/lib/python2.7/dist-packages/apache_beam/runners/portability/fn_api_runner.pyc in push(self, request)
   1059         request.instruction_id = 'control_%s' % self._uid_counter
   1060       logging.debug('CONTROL REQUEST %s', request)
-> 1061       response = self.worker.do_instruction(request)
   1062       logging.debug('CONTROL RESPONSE %s', response)
   1063       return ControlFuture(request.instruction_id, response)

/usr/local/lib/python2.7/dist-packages/apache_beam/runners/worker/sdk_worker.pyc in do_instruction(self, request)
    213       # E.g. if register is set, this will call self.register(request.register))
    214       return getattr(self, request_type)(getattr(request, request_type),
--> 215                                          request.instruction_id)
    216     else:
    217       raise NotImplementedError

/usr/local/lib/python2.7/dist-packages/apache_beam/runners/worker/sdk_worker.pyc in process_bundle(self, request, instruction_id)
    235     try:
    236       with state_handler.process_instruction_id(instruction_id):
--> 237         processor.process_bundle(instruction_id)
    238     finally:
    239       del self.bundle_processors[instruction_id]

/usr/local/lib/python2.7/dist-packages/apache_beam/runners/worker/bundle_processor.pyc in process_bundle(self, instruction_id)
    297             instruction_id, [input_op.target]):
    298           # ignores input name
--> 299           input_op.process_encoded(data.data)
    300 
    301       # Finish all operations.

/usr/local/lib/python2.7/dist-packages/apache_beam/runners/worker/bundle_processor.pyc in process_encoded(self, encoded_windowed_values)
    118       decoded_value = self.windowed_coder_impl.decode_from_stream(
    119           input_stream, True)
--> 120       self.output(decoded_value)
    121 
    122 

/usr/local/lib/python2.7/dist-packages/apache_beam/runners/worker/operations.so in apache_beam.runners.worker.operations.Operation.output()

/usr/local/lib/python2.7/dist-packages/apache_beam/runners/worker/operations.so in apache_beam.runners.worker.operations.Operation.output()

/usr/local/lib/python2.7/dist-packages/apache_beam/runners/worker/operations.so in apache_beam.runners.worker.operations.ConsumerSet.receive()

/usr/local/lib/python2.7/dist-packages/apache_beam/runners/worker/operations.so in apache_beam.runners.worker.operations.DoOperation.process()

/usr/local/lib/python2.7/dist-packages/apache_beam/runners/worker/operations.so in apache_beam.runners.worker.operations.DoOperation.process()

/usr/local/lib/python2.7/dist-packages/apache_beam/runners/common.so in apache_beam.runners.common.DoFnRunner.receive()

/usr/local/lib/python2.7/dist-packages/apache_beam/runners/common.so in apache_beam.runners.common.DoFnRunner.process()

/usr/local/lib/python2.7/dist-packages/apache_beam/runners/common.so in apache_beam.runners.common.DoFnRunner._reraise_augmented()

/usr/local/lib/python2.7/dist-packages/apache_beam/runners/common.so in apache_beam.runners.common.DoFnRunner.process()

/usr/local/lib/python2.7/dist-packages/apache_beam/runners/common.so in apache_beam.runners.common.PerWindowInvoker.invoke_process()

/usr/local/lib/python2.7/dist-packages/apache_beam/runners/common.so in apache_beam.runners.common.PerWindowInvoker._invoke_per_window()

/usr/local/lib/python2.7/dist-packages/apache_beam/io/iobase.pyc in process(self, element, init_result)
   1051     writer = self.sink.open_writer(init_result, str(uuid.uuid4()))
   1052     for e in bundle[1]:  # values
-> 1053       writer.write(e)
   1054     return [window.TimestampedValue(writer.close(), window.MAX_TIMESTAMP)]
   1055 

/usr/local/lib/python2.7/dist-packages/apache_beam/io/filebasedsink.pyc in write(self, value)
    386 
    387   def write(self, value):
--> 388     self.sink.write_record(self.temp_handle, value)
    389 
    390   def close(self):

/usr/local/lib/python2.7/dist-packages/apache_beam/io/filebasedsink.pyc in write_record(self, file_handle, value)
    135     this sink's Coder.
    136     """
--> 137     self.write_encoded_record(file_handle, self.coder.encode(value))
    138 
    139   def write_encoded_record(self, file_handle, encoded_value):

/usr/local/lib/python2.7/dist-packages/apache_beam/io/tfrecordio.pyc in write_encoded_record(self, file_handle, value)
    278 
    279   def write_encoded_record(self, file_handle, value):
--> 280     _TFRecordUtil.write_record(file_handle, value)
    281 
    282 

/usr/local/lib/python2.7/dist-packages/apache_beam/io/tfrecordio.pyc in write_record(cls, file_handle, value)
     99     file_handle.write('{}{}{}{}'.format(
    100         encoded_length,
--> 101         struct.pack('<I', cls._masked_crc32c(encoded_length)),  #
    102         value,
    103         struct.pack('<I', cls._masked_crc32c(value))))

/usr/local/lib/python2.7/dist-packages/apache_beam/io/tfrecordio.pyc in _masked_crc32c(cls, value, crc32c_fn)
     79     """
     80 
---> 81     crc = crc32c_fn(value)
     82     return (((crc >> 15) | (crc << 17)) + 0xa282ead8) & 0xffffffff
     83 

/usr/local/lib/python2.7/dist-packages/apache_beam/io/tfrecordio.pyc in _default_crc32c_fn(value)
     45     try:
     46       import snappy  # pylint: disable=import-error
---> 47       _default_crc32c_fn.fn = snappy._snappy._crc32c  # pylint: disable=protected-access
     48     except ImportError:
     49       logging.warning('Couldn\'t find python-snappy so the implementation of '

AttributeError: 'module' object has no attribute '_snappy' [while running 'WriteStatsOutput/Write/WriteImpl/WriteBundles']
```

However if I try 
`import snappy `
in a new notebook it loads properly.

EDIT: if I remove the package python-snappy the function works but with the following Warning:

> WARNING:root:Couldn't find python-snappy so the implementation of _TFRecordUtil._masked_crc32c is not as fast as it could be.

**Comment by paulgc** [comment]
Can you share which version of python-snappy and apache-beam you are using?

**Comment by vincentteyssier** [comment]
apache_beam v2.7.0
Python 2.7.12
pip 18.1
python-snappy 0.5-1build1

**Comment by paulgc** [comment]
The issue seems to be due to Apache Beam 2.7 not compatible with earlier versions of python-snappy. This should be resolved after https://github.com/apache/beam/pull/6801

**Comment by katsiapis** [comment]
The expectation is that the fix will make it to Apache Beam 2.9

/cc @aaltay

**Comment by harshini-gadige** [comment]
Closing as the PR mentioned above is merged.
